### PR TITLE
Introduce model module

### PR DESCRIPTION
**Pull request by zregvart**
This introduces the `github.com/redhat-appstudio/application-api/model` module containing the package version corresponding to the main API version and containing the bare struct types without any k8s dependencies. This is by design to minimize the downstream dependency tree and eliminate the dependency mismatch issues.

The `application.go` in `model/v1alpha1` is symlinked into `api/v1alpha1` to maintain backward compatibility.

Any downstream usage now can decide if it would depend on the `github.com/redhat-appstudio/application-api/` and inherit the dependency tree within it, or on `github.com/redhat-appstudio/application-api/module` without any dependencies. The later is most likely the best option for independently evolving the dependency tree.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,14 +5,17 @@ go 1.19
 require (
 	github.com/onsi/ginkgo v1.16.5
 	github.com/onsi/gomega v1.27.7
+	github.com/redhat-appstudio/application-api/model v0.0.0-00010101000000-000000000000
```
Comment (johnmcollier) [inline go.mod:8]: Trying to consume the changes here in HAS gives me the following error:
```
appstudio/application-api/model@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
```

Would changing version here to something like `v0.0.0` help?

**Review by johnmcollier — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (johnmcollier) [reply go.mod:8]: Actually, changing it to `v0.0.0` gives me a different error:
```
	github.com/redhat-appstudio/application-api/model: reading github.com/redhat-appstudio/application-api/model/go.mod at revision model/v0.0.0: unknown revision model/v0.0.0
```

**Review by johnmcollier — COMMENTED** [review]
Would this be considered a breaking change for existing consumers of the Application API? Would consumers need to update their code in order to work with the updated CRDs? Or would they continue to work?

File: go.mod (hunk 1 above)
Comment (zregvart) [reply go.mod:8]: There is no `github.com/redhat-appstudio/application-api/model`, well not unless this PR is merged.

This should work in a `go.mod`:

```
replace github.com/redhat-appstudio/application-api/model => github.com/zregvart/application-api/model v0.0.0-20230731114935-99992a8e0580

require github.com/redhat-appstudio/application-api/model v0.0.0-00010101000000-000000000000
```

**Comment by zregvart** [comment]
> Would this be considered a breaking change for existing consumers of the Application API? Would consumers need to update their code in order to work with the updated CRDs? Or would they continue to work?

Those that use the go API from this repository would find this to be a breaking change, yes. The option for them would be to remain on an older version of the go module until they're ready to make the change outlined above. The HTTP API should remain the same.

**Review by johnmcollier — COMMENTED** [review]
Looks good...

Before we merge, could you also revert the controller-runtime version change from https://github.com/redhat-appstudio/application-api/pull/58 so that the GitOps service becomes unblocked?

**Comment by zregvart** [comment]
We’re discussing a different approach in EC, most likely to vendor and sync the structs from here. I’ll convert this to draft until we see how that pans out.

**Comment by zregvart** [comment]
> Looks good...
> 
> Before we merge, could you also revert the controller-runtime version change from #58 so that the GitOps service becomes unblocked?

See #64

File: go.mod (hunk 1 above)
Comment (maysunfaisal) [reply go.mod:8]: I think this may also work

```
replace github.com/redhat-appstudio/application-api/model => ./model
```

Comment (yangcao77) [reply go.mod:8]: yes, so the following should work:

```
github.com/redhat-appstudio/application-api/model  v0.0.0


replace github.com/redhat-appstudio/application-api/model => ./model
```

**Comment by zregvart** [comment]
I've taken advantage of symlinks and now there should not be any downstream breaking code changes when using the go API. The new [description](https://github.com/redhat-appstudio/application-api/pull/63#issue-1828989017) now reflects that change.

**Comment by zregvart** [comment]
In df393ef2c1ccab636442ab7e7423df829f2c96d2 I've added the controller-runtime and k8s APIs downgrade to versions before #58

**Comment by zregvart** [comment]
Forgot that I need to do the same for other structs, will undraft when I push the updated version

**Comment by zregvart** [comment]
The Enterprise Contract team is trying the approach of vendoring select types from this repository in EC CLI. Follow [HACBS-2426](https://issues.redhat.com/browse/HACBS-2426) and [HACBS-2441](https://issues.redhat.com/browse/HACBS-2441) for details. We no longer require this, so I'll close. Thanks for the collaboration and reviews!

**Pull request closed without merging**